library/alloc/src/vec.rs

```diff
@@ -2188,14 +2188,16 @@ impl<T> Drop for InPlaceDrop<T> {
 
 impl<T> SpecFromIter<T, IntoIter<T>> for Vec<T> {
     fn from_iter(iterator: IntoIter<T>) -> Self {
-        // A common case is passing a vector into a function which immediately
-        // re-collects into a vector. We can short circuit this if the IntoIter
-        // has not been advanced at all.
-        // When it has been advanced We can also reuse the memory and move the data to the front.
-        // But we only do so when the resulting Vec wouldn't have more unused capacity
-        // than creating it through the generic FromIterator implementation would. That limitation
-        // is not strictly necessary as Vec's allocation behavior is intentionally unspecified.
-        // But it is a conservative choice.
+        // It is common to pass a vector to a function which immediately
+        // re-collects as a vector. We can short circuit this if the IntoIter
+        // has not been advanced.
+        //
+        // If it has been advanced, we can reuse the memory by moving the
+        // data to the front. But we only do so when the resulting Vec
+        // would have less unused capacity than extending a new Vec with the
+        // generic from_iter would. This limitation is not strictly necessary
+        // as Vec's allocation is intentionally unspecified.
+        // This is a conservative choice.
```

Comment:
I have no issue with rewrapping things, but it's not clear to me what the issue with the previous wording was. For example you're removing an "also" which is meant to say that the previous statement ("if ... has not been advanced") is not binding here anymore because we're applying an additional strategy.

Similarly the removed "but", which is meant to explain why we're doing this despite saying "is not strictly necessary". To me just saying "this" disconnects it from the chain of reasoning and describes the whole previous paragraph instead.

> than extending a new Vec with the generic from_iter would.

*Extending* might not be the best word here considering that this is not an `Extend` but a `FromIterator` implementation. Perhaps *growing*  would be a better choice.

Reply:
> I have no issue with rewrapping things, but it's not clear to me what the issue with the previous wording was. For example you're removing an "also" which is meant to say that the previous statement ("if ... has not been advanced") is not binding here anymore because we're applying an additional strategy.

The reason I removed "also" is because I believe it is uncessary, "we only do so" duplicates the meaning of also, do you think it is bad to remove "also"?

> Similarly the removed "but", which is meant to explain why we're doing this despite saying "is not strictly necessary". To me just saying "this" disconnects it from the chain of reasoning and describes the whole previous paragraph instead.

The reason why I do so is that I find the "but" thing weird there, it essentially means this whole choice so I replace "but it" with "this". Maybe I should say "this choice is conservative" or should I revert this part?

> Extending might not be the best word here considering that this is not an Extend but a FromIterator implementation. Perhaps growing would be a better choice.

I am finding a word for that, yeah I think growing would be easier to understand.

Reply:
> "we only do so" duplicates the meaning of also

The "also" provides an extension (shifting an advanced IntoIter) to the initial optimization (non-advanced IntoIter). The "we only do so" in turn applies a restriction (size/capacity ratio) on that extension but does not affect the initial case.

> The reason why I do so is that I find the "but" thing weird there, it essentially means this whole choice so I replace "but it" with "this".

No, it only applies to the restriction, not the extension.

Reply:
I still don't quite get it but I will just apply the change if you think it would be better since it looks the same to me.

Reply:
Look at the code. There are 3 cases (2 optimizations)

1. IntoIter has not been advanced - first paragraph, optimization 1
1. IntoIter has been advanced, so we move the data to the front. the "also" case. This optimization by itself would not be conservative. So saying "This" which might cover the optimization itself is inappropriate.
1. IntoIter has been advanced and the capacity is large compared to its case. we fall back to collecting into a new vec. This is the "but" and "we only do so" restriction on 2. and the conservative choice.

Since there is some confusion it seems like a good idea to rephrase it but I'm not sure where exactly the confusion lies and what I could improve.

Reply:
Let me work on this, I am typing the code out one by one in another project, at least I can try not to put everything in one chunk.

Reply:
```rust
    fn from_iter(iterator: IntoIter<T>) -> Self {
        // It is common to pass a vector to a function which immediately
        // re-collects as a vector. We can short circuit this behavior if
        // the IntoIter has not been advanced.
        let has_advanced = iterator.buf.as_ptr() as *const _ != iterator.ptr;
        if !has_advanced || iterator.len() >= iterator.cap / 2 {
            unsafe {
                let it = ManuallyDrop::new(iterator);
                // If it has been advanced, we reuse the memory by moving
                // the data to the front.
                if has_advanced {
                    ptr::copy(it.ptr, it.buf.as_ptr(), it.len());
                }
                return Vec::from_raw_parts(it.buf.as_ptr(), it.len(), it.cap);
            }
        }

        // A conservative choice to grow a new Vec with the generic from_iter
        // if there is a large unused capacity after the optimization above.
        // This limitation is not strictly necessary as Vec's allocation is
        // intentionally unspecified.
        let mut vec = Vec::new();
        // must delegate to spec_extend() since extend() itself delegates to spec_from for empty
        // Vecs
        vec.spec_extend(iterator);
        vec
    }
```
What do you think?

Reply:
Splitting it up into multiple places seems fine, but...

> if the IntoIter has not been advanced.

I think this can be a bit confusing when we do have special handling  for advanced iters inside the `if`.
The original implementation of this specialization only handled the non-advanced case and that's where the first paragraph of the comment cmes from. The `ptr::copy` is my improvement here.
If we restructure the comments then the first one should be updated to be more general to cover both optimization cases.

>  A conservative choice to grow a new Vec with the generic from_iter

Something about that phrasing still feels clunky.

Reply:
@the8472 Any suggestion to make it better? I tried my best already.

Reply:
I think I would have to rewrite the whole thing for it to make sense. The issue is that some of the comments old, stemming from the original optimization which I extended. That still made sense when immediately following up the original comment with the "also" extension. With them spread further apart the original comment is not appropriate anymore since it is too narrow.

Reply:
How about this?
```rust
    fn from_iter(iterator: IntoIter<T>) -> Self {
        // It is common to pass a vector to a function which immediately
        // re-collects as a vector. We can short circuit this behavior if
        // the IntoIter has not been advanced.
        //
        // We reuse the original buffer if the iterator has not been advanced
        // or if there is small unused capacity if we reuse the buffer.
        let has_advanced = iterator.buf.as_ptr() as *const _ != iterator.ptr;
        if !has_advanced || iterator.len() >= iterator.cap / 2 {
            unsafe {
                let it = ManuallyDrop::new(iterator);
                // If it has been advanced, we reuse the memory by moving
                // the data to the front.
                if has_advanced {
                    ptr::copy(it.ptr, it.buf.as_ptr(), it.len());
                }
                return Vec::from_raw_parts(it.buf.as_ptr(), it.len(), it.cap);
            }
        }

        // We use a conservative method to grow a new Vec with the generic
        // from_iter if there is a large unused capacity after the
        // optimization above. This limitation is not strictly necessary as
        // Vec's allocation is intentionally unspecified.
        let mut vec = Vec::new();
        // must delegate to spec_extend() since extend() itself delegates to spec_from for empty
        // Vecs
        vec.spec_extend(iterator);
        vec
    }
```
While writing the documentation, `iterator.len() >= iterator.cap / 2` does seem to look weird to me, looks kinda magical and I wonder if that works with both large and small capacity. Is there any reason why we are not comparing it with size hint too?

Reply:
Looks good now.

> While writing the documentation, iterator.len() >= iterator.cap / 2 does seem to look weird to me, looks kinda magical 

It mirrors the logic inside `RawVec` but there's no fitting method exposing it, so it's done manually here. 

> Is there any reason why we are not comparing it with size hint too?

`ExactSizeIterator::len()` is, well, exact.